### ISSUE TITLE
VS Code Ruby Debugger support

### ISSUE THREAD

**Issue by rebornix** [HookyQR/VSCodeRuby#1]
Hi HookyQR,
  This is Peng from MS. I noticed that you published this amazing plugin today, it's a really great trail on ruby integration. From the TODO list, it seems you are working on the ruby debug feature, it interests me as I'm also playing with VS Code ruby debugging feature these days and sent out the first prototype to the marketplace this weekend https://marketplace.visualstudio.com/items?itemName=rebornix.Ruby . Before that there were only a few basic Ruby intellisense or theming support in the marketplace. 
  I'm writing this issue to ask, is there any chance that we work together on this to avoid confusing users with multi plugins? You can get a rough idea about how much I have implemented from my repo https://github.com/rebornix/vscode-ruby .
  Looking forward to your response.

**Comment by HookyQR** [comment]
Hi Peng.

First off, I'd love to work together on this. 

I have a _semi_ working debug version (with very messy code - really was hacking at it) which used [ByeBug](https://github.com/deivid-rodriguez/byebug) as the back end, but found it to be difficult working with the output byebug provided. There was no separation between the executable's output and the debugger's output. I see from your example screen shot that you're not handling the STDOUT/STDERR from the executable at all.

I took a look at debug-ide as well, but didn't look into the split between the the two streams. Given that Ruby2 includes core capabilities for tracing and code inspection, I think it would be better on both ends (Ruby <-> JS) to write a new debugger element on the Ruby side that produces exactly the things that are required, but in JSON output for the debuggers comms, via a HTTP service, and still provides the executable output via STDOUT/STDERR.

I think this approach will give us the best user interface capability for VS Code, and 'extra' functionality (such as multi thread handling, and conditional breakpoints) should be pluggable as the extension grows.

This method will align with the necessary components for providing code completion as well.

Thoughts?

BWT. I noticed your extension just after I published mine when one of my other plugins told me mine was out of date!! Hence the change to [Extension Update Check](https://marketplace.visualstudio.com/items?itemName=HookyQR.ExtensionUpdateCheck). :wink:

**Comment by rebornix** [comment]
You are correct, I’m not handling stdin/out stream of ruby debug directly. Actually I leverage a gem called ruby debug ide https://github.com/ruby-debug/ruby-debug-ide to deal with it. This gem wraps the raw stream and convert debugging output to XML for later processing. 

AFAIK, RubyMine will install this gem for you for debugging while installation, that’s why I choose the gem as the debuuger interface from ruby side. It’s somewhat limited but it supports most of default debugging commands, like conditional breakpoint, threading. So I suppose it can be our last thing to replace it with our own ruby debug gem.

For now it works perfectly with existing VS Code debug functionality. My next move is finishing some missing part of stack trace and variables requests, then I will bring the debugger with threading support. You can take a look at my source code and trello board for more info  https://trello.com/b/5n80ssy7/vscode-ruby . We’ll see if my code base can be a good starting point sooner or later.

**Comment by HookyQR** [comment]
I had a look at your code, and it seems fine. I would expect the executable output to show in the debug output window though.

Does `rdebug-ide` give you the stdout stream separately? If so, then that should be fine. Input is another issue, but should be simple to solve: send to STDIN if the debugger is in a run state, and use it as variable lookup/debugger control if it's in a paused or exception state.

I'll have a deeper look at your code and the rdebug-ide interface tonight.

**Comment by rebornix** [comment]
Executable output and debugging output are sent through different IO connection, the former uses stdout while the latter uses stderr. For now I just send all the output to debug window but I can hide all debugging messages real quick.

I've just published v0.1.3 and you'll just see executable output from the debug window.

**Comment by HookyQR** [comment]
I had a run at your code last night and ran into some problems when adding multiple break points. It's currently in a non-working state, but I have made some excellent progress. I'll finish off what I have and push it up tonight. Do you want me to add the debugger to this one, or add a pull request to yours?

There are some substantial changes because of the response structure (mainly to do with multiple responses in one packet).

Looks like using rdebug-ide is good though. With the -xd flags off none of the debugger output taints the real output, so the exe flow can be shown without issue.

**Comment by rebornix** [comment]
@HookyQR there are still some bugs in ruby script debugging, I've made some fixes and updated the bits tonight. It would be great if you can make a pull request to my repo, once the code base is mature enough I suppose we can create an organization to maintain the code.

**Comment by HookyQR** [comment]
Seems you pushed after my last fork. I'll take another look tonight. My code is in the pull req. if you want to look.

**Comment by rebornix** [comment]
I've already looked at your code, the logic part is okay. The only comment
for now is maybe we can separate coding style change (like spacing,
brackets) with logic change, then the pr can be smaller can cleaner :)

Really appreciate your contrubition! My last push moves all ruby debugger
related code to a separate class and decouple it with VS Code DebugSession,
I suppose you change would be operated on `src/ruby. rb`.
